File: src/main/java/com/leijendary/spring/boot/template/core/generator/SnowflakeIdGenerator.java
package com.leijendary.spring.boot.template.core.generator;

import java.io.Serializable;

import com.leijendary.spring.boot.template.core.worker.SnowflakeIdWorker;

import org.hibernate.HibernateException;
import org.hibernate.engine.spi.SharedSessionContractImplementor;
import org.hibernate.id.IdentifierGenerator;

public class SnowflakeIdGenerator implements IdentifierGenerator {

    public static final String STRATEGY = "com.leijendary.spring.boot.template.generator.SnowflakeIdGenerator";

    private static final SnowflakeIdWorker WORKER = new SnowflakeIdWorker();

    @Override
    public Serializable generate(final SharedSessionContractImplementor session, final Object object)
            throws HibernateException {
        return WORKER.nextId();
    }
}

File: src/main/kotlin/com/leijendary/spring/boot/template/core/generator/SnowflakeIdGenerator.kt
package com.leijendary.spring.boot.template.core.generator

import com.leijendary.spring.boot.template.core.worker.SnowflakeIdWorker
import org.hibernate.HibernateException
import org.hibernate.engine.spi.SharedSessionContractImplementor
import org.hibernate.id.IdentifierGenerator
import java.io.Serializable

class SnowflakeIdGenerator : IdentifierGenerator {
    companion object {
        const val STRATEGY = "com.leijendary.spring.boot.template.generator.SnowflakeIdGenerator"
        private val WORKER: SnowflakeIdWorker = SnowflakeIdWorker()
    }

    @Throws(HibernateException::class)
    override fun generate(session: SharedSessionContractImplementor, any: Any): Serializable {
        return WORKER.nextId()
    }
}